changefeedccl: random changefeed frontier placement

stevendanna opened the pull request:
By default, CHANGEFEED jobs are adopted by the SQL instance that ran the SQL statement.

Further, we places the change frontier processor on the job coordinator node in part so that we could be 
sure that we were only updating the job record from the node that had the job claim.

However, this historically has caused hot-spots on nodes that found themselves as the coordinators
of a large number of changefeed jobs since the change aggregator does do a non-trivial amount of work.

Here, we randomly distribute the change aggregator to one of the eligible nodes.

To guard against concurrent job progress updates, we pass the claim session ID through the processor spec
so that we can still check the session on any update.

Release note: None
Epic: None

stevendanna committed:
changefeedccl: random changefeed frontier placement

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119293)
<!-- Reviewable:end -->

stevendanna commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1234,7 +1234,7 @@ func (cf *changeFrontier) Start(ctx context.Context) {
 
 	if cf.spec.JobID != 0 {
-		job, err := cf.flowCtx.Cfg.JobRegistry.LoadClaimedJob(ctx, cf.spec.JobID)
+		job, err := cf.flowCtx.Cfg.JobRegistry.LoadJob(ctx, cf.spec.JobID)
```
Not updating the the job from a machine that doesn't have the lease has, in the past at least, been a nice layer of protection against confusion caused by two copies of the job updating it concurrently.

I don't think we should ship this with this in place.

Rather, we might pursue one of the following:

1) Send the current job's lease in the spec so that we can continue to check that lease here. It might take some updates to the job code to make that work well but they should be minor.

2) Send the status updates we want to do back to the coordinator and do them there.

I worry that (2) might defeat some of the point of this change and be a rather large change to how this job works, so I prefer (1).

dt replied:
yeah, I think I prefer (1) as well, since writing to the job record might be "work" and I would like all "work" to be able to be scheduled off the coordinator. 

I suppose (2) could do the frontier work and just send simple-to-apply writes back to the coordinator to commit for it, but that just seems complicated.

stevendanna committed:
jobs,changefeeds: pass session ID of coordinator to frontier

This allows us to update the job from the frontier without giving up
the session check on update.

Epic: none

stevendanna commented:
@rharding6373 This probably has some work left still. But I wonder what y'all think in general?

stevendanna committed:
changefeedccl: disallow random placement on core changefeed

Epic: none
Release note: None

stevendanna commented:
I think the major TODO here would be figuring out how to deal with the `localState` that was added in e8b075b44e6bea78c01c596ce4a2a54f3c057422.

rharding6373 reviewed (commented):
Have a couple comments/questions, but overall seems fine.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-:-NsBB1vtEYcom9Vn7QoR:bbhpiav)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @stevendanna)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 507 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-NsAm0F98WBQwPvrQK9Y:-NsAm0F98WBQwPvrQK9Z:b5dsgde) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ae8e493b66f9db3f323afacb0ef37b61986dfe4/pkg/ccl/changefeedccl/changefeed_dist.go#L507)):*
> ```Go
> 		}
> 
> 		frontierSQLInstance := dsp.GatewayID()
> ```

For changefeeds and bulk jobs, is the gateway the node that claimed the job? Just an educational question. I've never thought of the gateway node that way before.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 373 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-NsB8XOz1p2KPku_1Fa-:-NsB8XOz1p2KPku_1Fa0:b-1gfeh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d5247b7b6e49615f2c75be25ce4ee9d3c8a6c50/pkg/ccl/changefeedccl/changefeed_dist.go#L373)):*
> ```Go
> 		"random", "gateway").(string),
> 	map[int64]string{
> 		int64(defaultDistribution):        "gateway",
> ```

Did you mean `frontierOnGateway` and `frontierOnRandomNode` here?
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 530 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-NsB9funAg7CWhtt2VZ5:-NsB9funAg7CWhtt2VZ6:b-l1d0zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d5247b7b6e49615f2c75be25ce4ee9d3c8a6c50/pkg/ccl/changefeedccl/changefeed_dist.go#L530)):*
> ```Go
> 						err.Error(), frontierSQLInstance)
> 				} else {
> 					frontierSQLInstance = all[rng.Intn(len(all))].InstanceID
> ```

If the user specifies `frontierOnRandomNode`, then would they expect the frontier to be randomized every time the job pauses and resumes? Or just randomized the first time and resumed on the same node (unless something else happened like a node died and the was forced to resume elsewhere)? Maybe we need another mode to randomize always?

Thinking of this in relation to your comment about the localState cache, which would not provide any benefit if the frontier location was randomized every time.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
Thanks for the review.

I am not sure yet what to do about the local state cache. That might be better investigated by someone who knows those recent changes better.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-:-NsDocbhCsNZ-0QxIAYc:bfvfh2h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @rharding6373)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 507 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-NsAm0F98WBQwPvrQK9Y:-NsDovltAOZCvq3uinUO:beoqaot) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ae8e493b66f9db3f323afacb0ef37b61986dfe4/pkg/ccl/changefeedccl/changefeed_dist.go#L507)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

For changefeeds and bulk jobs, is the gateway the node that claimed the job? Just an educational question. I've never thought of the gateway node that way before.
</blockquote></details>

Yes, for DistSQL flows that are started from a Resumer, the gateway is the node running the resumer (i.e. the node that has claimed the job).
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 373 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-NsB8XOz1p2KPku_1Fa-:-NsDothcC5uhWjPWC_Hs:b8pelt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d5247b7b6e49615f2c75be25ce4ee9d3c8a6c50/pkg/ccl/changefeedccl/changefeed_dist.go#L373)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Did you mean `frontierOnGateway` and `frontierOnRandomNode` here?
</blockquote></details>

Oops. Thanks.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 530 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119293#-NsB9funAg7CWhtt2VZ5:-NsDpML-DXcM0pp0pEGY:b-rmhay7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d5247b7b6e49615f2c75be25ce4ee9d3c8a6c50/pkg/ccl/changefeedccl/changefeed_dist.go#L530)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

If the user specifies `frontierOnRandomNode`, then would they expect the frontier to be randomized every time the job pauses and resumes? Or just randomized the first time and resumed on the same node (unless something else happened like a node died and the was forced to resume elsewhere)? Maybe we need another mode to randomize always?

Thinking of this in relation to your comment about the localState cache, which would not provide any benefit if the frontier location was randomized every time.
</blockquote></details>

I'm unsure. Overall, I think users who want this mostly just care that overall their cluster doesn't have hotspots.  On PAUSE/RESUME I think that matters a little less since you get some randomization just from the job system adopting things across the whole cluster.  


<!-- Sent from Reviewable.io -->
